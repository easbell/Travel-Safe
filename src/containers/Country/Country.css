.country {
  display: flex;
  align-items: center;
  justify-content: space-between;
  box-shadow: -4px 4px 11px -5px rgba(97,97,97,1);
  width: 50%;
  margin: 40px auto;
  height: 110px;
  border-radius: 5px;
  cursor: pointer;
  padding: 0px 110px;
  background: #fff;
  opacity: .8;
  color: #000;
}

.country:hover {
  box-shadow: -7px 7px 12px -5px rgba(97,97,97,1);
  opacity: 1;
}

.rating {
  border-radius: 50%;
  height: 55px;
  width: 55px;
  font-size: 22px;
  color: #fff;
  border: 0;
  text-decoration: none;
}

.L {
  background-color: #37d57b;
}

.M {
  background-color: #82b8f7;
}

.H {
  background-color: #fbf282;
}

.E {
  background-color: #ee3649;
}