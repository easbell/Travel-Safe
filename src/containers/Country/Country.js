import React, { Component } from 'react';
import classnames from 'classnames';
import './Country.css';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import cogoToast from 'cogo-toast';

export class Country extends Component {
  constructor() {
    super();
    this.state = {
      rating: '',
      saved: false
    }
  }

  componentDidMount() {
    this.checkRating();
    this.checkSaved();
  }

  checkSaved = ()  => {
    const { id } = this.props
    const localSaved = localStorage.getItem('saved')
    if(localSaved && localSaved.includes(id)) {
      this.setState({ saved: true })
    }
  }

  checkRating = () => {
    const { rating } = this.props;
    if(rating <= 2.5) {
      this.setState({rating: 'L'})
    } else if(2.6 <= rating && rating <= 3.5) {
      this.setState({rating: 'M'})
    } else if(3.6 <= rating && rating <= 4.5) {
      this.setState({rating: 'H'})
    } else if(rating > 4.5) {
      this.setState({rating: 'E'})
    }
  }

  displayMessage = (message) => {
    cogoToast.success(message, { 
      position: 'bottom-left'
    });
  }

  addCountry = () => {
    const { id } = this.props;
    const localSaved = JSON.parse(localStorage.getItem('saved'))
    if(localSaved && !localSaved.includes(id)) {
      localStorage.setItem('saved', JSON.stringify([...localSaved, id]));
    } else if(!localSaved) {
      localStorage.setItem('saved', JSON.stringify([id]));
    }
    this.displayMessage('Country was added to favorites.')      
  }

  toggleCountry = (e) => {
    const { id } = this.props;
    if(e.target.className.includes('countrySaved')) {
      this.setState({ saved: false });
      const localSaved = JSON.parse(localStorage.getItem('saved'))
      const filteredSaved = localSaved.filter(country => country !== id)
      localStorage.setItem('saved', JSON.stringify(filteredSaved));
      this.displayMessage('Country removed from favorites.')
    } else {
      this.setState({ saved: true });
      this.addCountry();
    }
  }
  
  render() {
    const classNames = require('classnames');
    const { id, name } = this.props;
    const { rating, saved } = this.state;
    const btnClass = classNames({
      btn: true,
      add: true,
      countrySaved: saved
    });

    return (
      <div className='country'>
        {saved && <button className={btnClass} onClick={this.toggleCountry}>-</button>}
        {!saved && <button className={btnClass} onClick={this.toggleCountry}>+</button>}
        <Link to={`/details/${id}`} style={{ textDecoration: 'none' }}>
          <h2>{name}</h2>
        </Link>
        <button className={classnames('btn', rating)}>{rating}</button>
      </div>
    )
  }
}

export const mapStateToProps = state => ({
  saved: state.saved
})

export default connect(mapStateToProps)(Country);