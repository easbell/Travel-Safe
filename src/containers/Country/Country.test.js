import React from 'react';
import { Country } from './Country';
import { shallow } from 'enzyme';
import cogoToast from 'cogo-toast';

jest.mock('cogo-toast');

describe('Country', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Country 
        key={'France'} 
        name={'France'}
        rating={2.4}
    />)
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call checkRating when component is mounted', () => {
    const spy = jest.spyOn(wrapper.instance(), 'checkRating');
    wrapper.instance().componentDidMount();
    expect(spy).toHaveBeenCalled();
  });

  describe('checkRating', () => {
    it('should return rating L when checkRating is called', () => {
      wrapper.instance().checkRating();
      expect(wrapper.state('rating')).toEqual('L');
    });
  
    it('should return rating M when checkRating is called', () => {
      wrapper = shallow(<Country rating={3}/>);
      wrapper.instance().checkRating();
      expect(wrapper.state('rating')).toEqual('M');
    });
  
    it('should return rating H when checkRating is called', () => {
      wrapper = shallow(<Country rating={4}/>);
      wrapper.instance().checkRating();
      expect(wrapper.state('rating')).toEqual('H');
    });
  
    it('should return rating E when checkRating is called', () => {
      wrapper = shallow(<Country rating={5}/>);
      wrapper.instance().checkRating();
      expect(wrapper.state('rating')).toEqual('E');
    });
  })

  describe('addCountry', () => {
    let localStorageMock;

    beforeEach(() => {
      localStorageMock = {
        getItem: jest.fn(),
        setItem: jest.fn()
      };
      global.localStorage  = localStorageMock;
    });

    it('should call localStorage.getItem', () => {
      wrapper = shallow(<Country id={5}/>);
      wrapper.find('.add').simulate('click');
      expect(localStorage.getItem).toHaveBeenCalled();
      expect(localStorage.setItem).toHaveBeenCalled();
    });

    it('calls cogoToast.success', () => {
      cogoToast.success = jest.fn();
      wrapper = shallow(<Country id={5}/>);
      wrapper.find('.add').simulate('click');
      expect(cogoToast.success).toHaveBeenCalled();
    });
  });
})