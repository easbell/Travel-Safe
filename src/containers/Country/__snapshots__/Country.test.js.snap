// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Country should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Country
    name="France"
    rating={2.4}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        <button
          className="btn add"
          onClick={[Function]}
        >
          +
        </button>,
        <Link
          style={
            Object {
              "textDecoration": "none",
            }
          }
          to="/details/undefined"
        >
          <h2>
            France
          </h2>
        </Link>,
        <button
          className="btn L"
        >
          L
        </button>,
      ],
      "className": "country",
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "+",
          "className": "btn add",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "+",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <h2>
            France
          </h2>,
          "style": Object {
            "textDecoration": "none",
          },
          "to": "/details/undefined",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "France",
          },
          "ref": null,
          "rendered": "France",
          "type": "h2",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "L",
          "className": "btn L",
        },
        "ref": null,
        "rendered": "L",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <button
            className="btn add"
            onClick={[Function]}
          >
            +
          </button>,
          <Link
            style={
              Object {
                "textDecoration": "none",
              }
            }
            to="/details/undefined"
          >
            <h2>
              France
            </h2>
          </Link>,
          <button
            className="btn L"
          >
            L
          </button>,
        ],
        "className": "country",
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "+",
            "className": "btn add",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "+",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <h2>
              France
            </h2>,
            "style": Object {
              "textDecoration": "none",
            },
            "to": "/details/undefined",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "France",
            },
            "ref": null,
            "rendered": "France",
            "type": "h2",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "L",
            "className": "btn L",
          },
          "ref": null,
          "rendered": "L",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
